agent/dockerclient/dockerauth/asm_test.go

```diff
@@ -0,0 +1,121 @@
+package dockerauth
+
+import (
+	"testing"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/async"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	region            = "us-west-2"
+	authDataCacheSize = 100
+	authDataCacheTTL  = 12 * time.Hour
+)
+
+func TestNewAuthProviderASMAuth(t *testing.T) {
+
+	registryAuthData := &api.RegistryAuthenticationData{
+		Type: "asm",
+		ASMAuthData: &api.ASMAuthData{
+			Region:               "us-west-2",
+			CredentialsParameter: "secrets-value-arn",
+		},
+	}
+
+	provider := NewASMAuthProvider(registryAuthData, async.NewLRUCache(authDataCacheSize, authDataCacheTTL))
+	_, ok := provider.(*asmAuthProvider)
+	assert.True(t, ok, "Should have returned asmAuthProvider")
+}
+
+type mockGetSecretValue struct {
+	secretsmanageriface.SecretsManagerAPI
+	Resp secretsmanager.GetSecretValueOutput
+}
+
+func (m mockGetSecretValue) GetSecretValue(input *secretsmanager.GetSecretValueInput) (*secretsmanager.GetSecretValueOutput, error) {
+	return &m.Resp, nil
+}
+
+func TestASMGetAuthConfig(t *testing.T) {
+
+	registryAuthData := &api.RegistryAuthenticationData{
+		Type: "asm",
+		ASMAuthData: &api.ASMAuthData{
+			CredentialsParameter: "secrets-value-arn",
+		},
+	}
+
+	registryAuthData.ASMAuthData.SetTaskARN("arn")
+
+	cases := []struct {
+		Data        *api.RegistryAuthenticationData
+		Resp        secretsmanager.GetSecretValueOutput
+		ShouldError bool
+	}{
+		{ //case 1, success
+			Data: registryAuthData,
+			Resp: secretsmanager.GetSecretValueOutput{
+				SecretString: aws.String(`{"username":"usr","password":"pwd"}`),
+			},
+			ShouldError: false,
+		},
+		{ //case 2, missing username
+			Data: registryAuthData,
+			Resp: secretsmanager.GetSecretValueOutput{
+				SecretString: aws.String(`{"password":"pwd"}`),
+			},
+			ShouldError: true,
+		},
+		{ //case 3, missing password
+			Data: registryAuthData,
+			Resp: secretsmanager.GetSecretValueOutput{
+				SecretString: aws.String(`{"username":"usr"}`),
+			},
+			ShouldError: true,
+		},
+		{ //case 4, malformed json
+			Data: registryAuthData,
+			Resp: secretsmanager.GetSecretValueOutput{
+				SecretString: aws.String(`{"username":"usr"`),
+			},
+			ShouldError: true,
+		},
+		{ //case 5, missing SecretString
+			Data:        registryAuthData,
+			Resp:        secretsmanager.GetSecretValueOutput{},
+			ShouldError: true,
+		},
+		{ //case 6, missing ASMAuthData
+			Data:        &api.RegistryAuthenticationData{Type: "asm"},
+			Resp:        secretsmanager.GetSecretValueOutput{},
+			ShouldError: true,
+		},
+		{ //case 7, missing RegistryAuthenticationData
```

Comment:
nit:
make the first arg `Name` and use that instead of a comment.

Then, in the for loop use subtests:

```
for _, c := range cases {
    t.Run(c.Name, func(t *testing.T) {
        // put the rest of your test logic in here
    })
}
```

Then your comment / case name will be exposed in the message if your test fails.

Reply:
this is helpful, thank you.